feat: pool escrow

wischli opened the pull request:
Adds per pool escrow(s) that internally separate by share class ids
* Deployed via new factory `PoolEscrowFactory` during `poolManager.addShareClass`
* Exposes deterministic retrieval of escrow address for a given pool id (`poolEscrowFactory.escrow(poolId)`) which is more lightweight (roughly 600-900 gas) than a storage read (2100-2600 gas).

## Backwards compatibility with Legacy v2 vaults

* Legacy vaults retrieve the global escrow from the async investment manager `manager.escrow()` ([ref](https://github.com/centrifuge/vaults-internal/blob/5e8262b26f8b3b488fa11855bfc2ee2e9943c09c/src/ERC7540Vault.sol#L102))
* Therefore, the previous implicit `BaseInvestmentManager.escrow()` from the public escrow storage was replaced with a custom function that assumes the message sender to be `vault` such that we can derive the pool id via `IBaseVault(vault).poolId()`. We do this using `staticcall` which reverts if the `msg.sender` does not support `poolId()`
* Nevertheless, we use `poolEscrowProvider.escrow(poolId)` in v3 vaults to save gas due to not requiring calling into the underlying investment manager 

TODOs

- [x] Add tests for escrow factory 
- [x] Ensure solution is lightweight
- [x] Ensure not breaking for legacy vaults + add docs for that

wischli committed:
feat: pool escrow

wischli commented on src/hooks/RestrictedTransfers.sol:
```diff
@@ -67,13 +70,13 @@ contract RestrictedTransfers is Auth, IRestrictedTransfers, IHook {
 
-        if (root.endorsed(to) || to == address(0)) {
-            // Destination is endorsed and source was already checked, so the transfer is allowed
+        if (root.endorsed(to) || to == address(0) || to == ESCROW_IDENTIFIER) {
```
NOTE: These changes were necessary because previously the global escrow contract was endorsed by root during deployment which we cannot do anymore as escrows are deployed dynamically by the corresponding factory.

wischli committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

wischli committed:
tests: add escrow factory

wischli committed:
tests: full escrow coverage

wischli committed:
feat: creep view onERC20Transfer

wischli commented on test/vaults/mocks/MockHook.sol:
```diff
@@ -9,4 +9,6 @@ contract MockHook is Mock {
         return interfaceId == type(IHook).interfaceId || interfaceId == type(IERC165).interfaceId;
     }
+
+    function updateRestriction(address token, bytes memory update) external {}
```
NOTE: Required for `PoolManager` test which just needs to be able to call it.

wischli committed:
chore: cleanup + legacy support

wischli committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

wischli committed:
fix: backwards escrow compatibility

wischli committed:
tests: improve escrow factory

wischli committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

wischli committed:
tests: block legacy pool for deploy vault tests

wischli committed:
tests: fix BalanceSheet missing ward checks

hieronx commented on src/vaults/PoolManager.sol:
```diff
@@ -210,7 +203,18 @@ contract PoolManager is Auth, Recoverable, IPoolManager, IUpdateContract, IPoolM
+        }
+
+        // Update escrow as member to enable minting, burning and transfers on deposit and redeem
+        bytes memory updateMember =
```
This won't apply to every hook, since some might not support the `UpdateRestrictionMember` message.

Also some might not have a hook at all (`address(0)`), for fully permissionless hooks.

I think it is just up to the pool admin to ensure this message is called. Our UI can batch it.

hieronx commented on src/vaults/SyncRequests.sol:
```diff
@@ -315,7 +317,13 @@ contract SyncRequests is BaseInvestmentManager, ISyncRequests {
+            scId,
+            vaultDetails.asset,
+            vaultDetails.tokenId,
+            address(poolEscrowProvider.poolEscrow(poolId.raw())),
```
Nit: if you put `address escrow =  address(poolEscrowProvider.poolEscrow(poolId.raw()))` on the line above this, I believe the rest of the call would fit on 1 line.

hieronx commented on src/hooks/FreelyTransferable.sol:
```diff
@@ -32,13 +32,14 @@ contract FreelyTransferable is Auth, IRestrictedTransfers, IHook {
     uint8 public constant FREEZE_BIT = 0;
+    /// @dev Magic address denoting a transfer to the escrow
+    /// @dev Solely used for gas saving since escrow is per pool
+    address public constant ESCROW_IDENTIFIER = address(uint160(uint8(0xce)));
```
I'd add this to `IHook.sol`, since that is the shared interface for hooks

wischli replied:
[5991500](https://github.com/centrifuge/protocol-v3/pull/262/commits/599150076de04b9aa97c6729d096fba0c5d9800a)

hieronx commented on src/vaults/BaseVaults.sol:
```diff
@@ -178,6 +188,12 @@ abstract contract BaseVault is Auth, Recoverable, IBaseVault {
 
+    // --- Legacy views ---
+    /// @inheritdoc IBaseVault
+    function escrow() public view returns (address) {
```
Doesn't this call need to be in `BaseInvestmentManager`? Since the v2.1 vaults call `manager.escrow()`

hieronx commented on src/hooks/FreelyTransferable.sol:
```diff
@@ -32,19 +32,21 @@ contract FreelyTransferable is Auth, IRestrictedTransfers, IHook {
     /// @inheritdoc IHook
     function onERC20Transfer(address from, address to, uint256 value, HookData calldata hookData)
         external
+        view
```
I don't think we should do this, see reasoning on Discord

hieronx commented on src/vaults/factories/PoolEscrowFactory.sol:
```diff
@@ -0,0 +1,85 @@
+    /// @inheritdoc IEscrowProvider
+    function escrow(uint64 poolId) external view returns (address) {
+        if (poolId == V2_POOL_ID) {
+            return V2_GLOBAL_ESCROW;
```
I likely am missing something: why are we still keeping the old global escrow around? I believe we can migrate this entirely.

wischli replied:
Do we need to support vaults from more than one pool id? If so, the `BaseInvestmentManager.escrow()` would be lacking the pool id information such that it can only map to one new global v2 escrow. Otherwise, good point that we could just migrate it.

hieronx replied:
> would be lacking the pool id information such that it can only map to one new global v2 escrow

It can just use `msg.sender` to look up the pool ID based on the vault address.

wischli replied:
[d72ad0a](https://github.com/centrifuge/protocol-v3/pull/262/commits/d72ad0a1e9fe0998e58731d1f0ea155a38fc0835)

lemunozm commented on src/vaults/factories/PoolEscrowFactory.sol:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /// @inheritdoc IPoolEscrowFactory
+    function newEscrow(uint64 poolId) public auth returns (address) {
```
NIT. Could we return here a `IPoolEscrow`?

wischli replied:
Due to the refactor in [d72ad0a](https://github.com/centrifuge/protocol-v3/pull/262/commits/d72ad0a1e9fe0998e58731d1f0ea155a38fc0835) which removed `poolEscrow` and `IEscrowProvider`, I decided to just return the address and require consumers to type cast if necessary.

lemunozm replied:
Ok. Just to notify, if you return an address instead of `IPoolEscrow` to avoid breaking retro-compatibility, both types, AFAIK, have the same abi. 

Is there any reason to keep the `address` as type when it always is a `IPoolEscrow` interface?

wischli replied:
[99cb82a](https://github.com/centrifuge/protocol-v3/pull/262/commits/99cb82a42bf8505c80782b08cfcd664d803b9fd6)

lemunozm reviewed (commented):
I couple of NITs, but I need still to review this

wischli committed:
fix: legacy escrow migration

wischli committed:
ref: hook escrow id

wischli committed:
fix: remove update member escrow in addShareClass

wischli commented on src/vaults/AsyncRequests.sol:
```diff
@@ -45,21 +47,26 @@ contract AsyncRequests is BaseInvestmentManager, IAsyncRequests {
 
+    /// @dev Magic address denoting a transfer to the escrow
+    /// @dev Solely used for gas saving since escrow is per pool
+    address public constant ESCROW_IDENTIFIER = address(uint160(uint8(0xce)));
```
[5991500](https://github.com/centrifuge/protocol-v3/pull/262/commits/599150076de04b9aa97c6729d096fba0c5d9800a)

wischli committed:
fix: revert added view on FreelyTransferable.onERC20Transfer

lemunozm commented on src/vaults/BaseInvestmentManager.sol:
```diff
@@ -67,6 +68,17 @@ abstract contract BaseInvestmentManager is Auth, Recoverable, IBaseInvestmentMan
     }
 
+    /// @inheritdoc IBaseInvestmentManager
+    function escrow() public view returns (address) {
+        (bool success, bytes memory result) = msg.sender.staticcall(abi.encodeWithSelector(IBaseVault.poolId.selector));
+
+        require(success && result.length == 32, SenderNotVault());
+
+        uint64 poolId = abi.decode(result, (uint64));
+        return poolEscrowProvider.escrow(poolId);
+    }
```
Really smart solution! 🙌🏻

mustermeiszer replied:
Thats really smart! Great solution

lemunozm commented on src/vaults/AsyncVault.sol:
```diff
@@ -46,9 +51,9 @@ contract AsyncVault is AsyncRedeemVault, IAsyncVault {
 
         if (tokenId == 0) {
-            SafeTransferLib.safeTransferFrom(asset, owner, asyncManager().escrow(), assets);
+            SafeTransferLib.safeTransferFrom(asset, owner, _poolEscrowProvider.escrow(poolId), assets);
```
Ok, I understand we do not break compatibility because `BaseInvestmentManager.escrow()` still exists.

wischli committed:
tests: apply new escrow logic to recon

lemunozm reviewed (dismissed):
Everything looks amazing!!

Maybe we can add the link to the exact old deployed vault contract, to be able to understand the changes and in the future avoid breaking the old vaults. Not sure where to add this, maybe a retrocompatibility doc file. But not for now! Good job! 🥇

mustermeiszer commented on src/vaults/AsyncRequests.sol:
```diff
@@ -259,7 +265,7 @@ contract AsyncRequests is BaseInvestmentManager, IAsyncRequests {
         // Burn redeemed share class tokens from escrow
         IShareToken shareToken = IShareToken(IAsyncVault(vault_).share());
-        shareToken.burn(address(escrow), shares);
+        shareToken.burn(poolEscrowProvider.escrow(poolId), shares);
```
QUESTION

Won't this call into the `hook` too and then block as it is not he `ESCROW_HOOK_ID`?

wischli replied:
Yes, this is the reason why we have to add the pool escrow as a member, [see above](https://github.com/centrifuge/protocol-v3/pull/262#discussion_r2046338528). I removed doing that inside `PoolManagar.addShareClass`, because Jeroen pointed out that this should rather be done manually by the FM ([ref](https://github.com/centrifuge/protocol-v3/pull/262#discussion_r2047689497)).

Therefore, we do this in vault integration tests inside our mock `centrifugeChain.addShareClass`.

mustermeiszer replied:
Just a thought @hieronx couldn't somebody use the escrow as a provider of funds? IIRC this is possible as we use `transferFrom` in the `SyncReyest` contract, right? IN the `AsyncRequest` it is the vault, this one does not have permission, but in the `SyncRequest` it is the manager.

mustermeiszer replied:
CR

Also just checked, the given flow will not work with the current managers - as `balanceSheet.deposit()` expects the tokens to already be there -part of the `manager.mint()` flow , but the `BaseVault` only transfers them after `manager.mint()` call. 

we probably need to do change the logic in `baseVault.mint()` to 
1.  `manager.preMint()` - issuance and transfer os shares
2. `SafeTransferFrom()` - token to escrow
3.  `manager.mint()` - `balanceSheet.deposit()` logic

wischli replied:
Unfortunately, we cannot change the `mint()` behaviour of legacy vaults, so it appears to me that we need to find another solution?

Just a quick question: Since this flow wasn't changed, it appears to me that it was broken before as well?

wischli replied:
Side question after looking deeper into the sync mint flow: I don't fully understand our `BaseSyncDepositVault.mint`: When we execute `syncDepositManager.mint(address(this, shares, receiver, msg.sender))` we 
1. Mint shares to the receiver via `balanceSheet.issue`
2. Execute `balanceSheet.deposit()` with **zero asset amount**. I.e., we neither increase the pending deposit, nor transfer anything and also send a holding update with zero increased amount. However, we then do the actual asset transfer afterwards in `SafeTransferLib.safeTransferFrom(asset, msg.sender, address(_poolEscrowProvider.escrow(poolId)), assets)`.

So shouldn't `syncRequests.mint -> _issueShares` actually do 

```diff
        balanceSheet.deposit(
-            poolId, scId, vaultDetails.asset, vaultDetails.tokenId, escrow, depositAssetAmount = 0, priceData.poolPerAsset
+            poolId, scId, vaultDetails.asset, vaultDetails.tokenId, owner, depositAssetAmount > 0, priceData.poolPerAsset
        );
```

Where `owner` is the `msg.sender` of `BaseSyncDepositVault` such that we can also decrease the `BaseSyncDepositVault.mint` to

```diff
        assets = syncDepositManager.mint(address(this), shares, receiver, msg.sender);
-        SafeTransferLib.safeTransferFrom(asset, msg.sender, address(_poolEscrowProvider.escrow(poolId)), assets);
        emit Deposit(receiver, msg.sender, assets, shares);
    }
```

wischli replied:
@mustermeiszer Committed my proposal, curious about your thoughts: https://github.com/centrifuge/protocol-v3/pull/262/commits/b679255c000eee1d424d874cdff80de980911c6a

Not sure if this is acceptable (assuming it's not) because then the transfer of the assets is not initiated from the vault but the BSM such that it needs approval.

hieronx replied:
@wischli this doesn't work because with ERC-4626 `deposit`/`mint`, the approval of assets needs to be to the vault address (as that is the only address an integrator knows). With your alternative proposal, the ERC20 allowance would need to be given to the `BalanceSheet`.

hieronx replied:
> That is just a requirement for the sync vaults. So we can change that. The async ones work with pending and increase just fine

For re-entrancy protection it is actually better if the transfer is at the end of the transaction. See the discussions with Alex about the hook security considerations.

Maybe we have a ` bool externalTransfer` in `deposit()` which basically doesn't check the balance? Not super clean but seems like the safest approach.

hieronx replied:
Or we just add `protected` reentrancy guard to `syncDepositVault.deposit/mint`, maybe cleaner.

wischli replied:
Fixed in two commits:
1. [2f8d13b](https://github.com/centrifuge/protocol-v3/pull/262/commits/2f8d13ba54728e6c55507bcfe154edad662f3fc7): Adds `BalanceSheet.noteDeposit` but does not perform transfers at the end of the vault calls
2. [aa7ba3a](https://github.com/centrifuge/protocol-v3/pull/262/commits/aa7ba3abb83096af2e23e9b1caa7b0dcf80b10d6): Adds `Escrow.noteDeposit` such that we can execute transfers last

As discussed in the sync, this solution is hacky but I also couldn't come up with something better that achieves the same result this effectively.

mustermeiszer commented on src/vaults/AsyncRequests.sol:
```diff
@@ -230,7 +236,7 @@ contract AsyncRequests is BaseInvestmentManager, IAsyncRequests {
         // Mint to escrow. Recipient can claim by calling deposit / mint
         IShareToken shareToken = IShareToken(IAsyncVault(vault_).share());
-        shareToken.mint(address(escrow), shares);
+        shareToken.mint(poolEscrowProvider.escrow(poolId), shares);
```
QUESTION

Won't this call into the `hook` too and then block as it is not he `ESCROW_HOOK_ID`?

mustermeiszer commented on src/vaults/BaseVaults.sol:
```diff
@@ -87,6 +95,8 @@ abstract contract BaseVault is Auth, Recoverable, IBaseVault {
     // --- Administration ---
     function file(bytes32 what, address data) external auth {
         if (what == "manager") manager = IBaseInvestmentManager(data);
+        /// @dev NOT supported in legacy v2 vaults
```
❤️

mustermeiszer commented on src/vaults/factories/PoolEscrowFactory.sol:
```diff
@@ -0,0 +1,64 @@
+
+        escrow_.rely(root);
+        escrow_.rely(poolManager);
+        escrow_.rely(balanceSheet);
+        escrow_.rely(asyncRequests);
```
QUESTION

Do we need to rely them? Isn't the approval enough?

wischli replied:
We need to rely all of these contracts because they are executing escrow calls.

mustermeiszer replied:
Ah, yes! the increase and so on! Perfect!

mustermeiszer reviewed (commented):
Amazing work! 

Do we check that `mint/burn` work as the `CentrifugeToken` seems to check the `onTransfer` logic that calls into the hook. I would expect those to fail if we receiver or sender is the new pool escrow. Happy if it works. Just wanted to highlight that. 
If it fails, couldn't we store the escrow in the hook itself as a storage item? 

Added the requirement for updating the hook of legacy vaults [here](https://www.notion.so/Deployment-Legacy-1b92eac24e178042a6d1d1881807d717?pvs=4#1d82eac24e178087b0f9cf96c459eb02) to our list in notion.

wischli committed:
ref: explicit type return for escrow

wischli committed:
fix: sync deposit + mint

lemunozm commented on src/vaults/AsyncRequests.sol:
```diff
@@ -128,7 +128,7 @@ contract AsyncRequests is BaseInvestmentManager, IAsyncRequests {
 
         state.pendingDepositRequest += assets;
-        IPoolEscrow(poolEscrowProvider.escrow(poolId)).pendingDepositIncrease(
+        IPoolEscrow(address(poolEscrowProvider.escrow(poolId))).pendingDepositIncrease(
```
NIT. THis can be:
```suggestion
        poolEscrowProvider.escrow(poolId).pendingDepositIncrease(
```
If poolEscrowProvider already store IPoolEscrow in the mapping

lemunozm committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

lemunozm committed:
Feat: Refund per escrow (#266)

* base implementation

minor change

fix permission issue

implement new solution

fixes

minor adition

fix test

fix gas snapshot

fix tests

* update/fix recover system

* minor test addition

* Simplify escrow (#275)

* base implementation

* minor change

* fix permission issue

* implement new solution

* fixes

* minor adition

* fix test

* fix gas snapshot

* fix tests

* Try to simplify, add notes

* Remove redundant comment

* Some consistency changes

* Undo

* Fix events

---------

Co-authored-by: lemunozm <lemunozm@gmail.com>

---------

Co-authored-by: Jeroen <1748621+hieronx@users.noreply.github.com>

mustermeiszer commented on src/vaults/AsyncRequests.sol:
```diff
@@ -150,8 +153,8 @@ contract AsyncRequests is BaseInvestmentManager, IAsyncRequests {
-            _canTransfer(vaultAddr, owner, address(escrow), shares)
-                && _canTransfer(vaultAddr, controller, address(escrow), shares),
+            _canTransfer(vaultAddr, owner, ESCROW_HOOK_ID, shares)
+                && _canTransfer(vaultAddr, controller, ESCROW_HOOK_ID, shares),
```
Just a side note: 

* IF we anuyways need to whitelist the escrow in order for this too work, why have the `ESCROW_HOOK_ID` at all?

wischli replied:
Seems feasible to me. @hieronx WDYT?

hieronx replied:
To be able to identify redemption requests (https://github.com/centrifuge/protocol-v3/blob/main/src/vaults/interfaces/token/IHook.sol#L21)

wischli committed:
fix: sync vault deposit + mint

wischli committed:
ref: minor rename

wischli committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

wischli committed:
ref: rm legacy comments + BaseInvestmentManager.escrow

lemunozm commented:
Maybe we can merge this already to avoid future conflicts with others refactorings (?)

wischli committed:
fix: transfer last

wischli committed:
Merge remote-tracking branch 'origin/main' into ref/per-pool-escrow

mustermeiszer commented on src/vaults/BalanceSheet.sol:
```diff
@@ -229,19 +245,26 @@ contract BalanceSheet is Auth, Recoverable, IBalanceSheet, IBalanceSheetGatewayH
         address provider,
         uint128 amount,
-        D18 pricePoolPerAsset
+        D18 pricePoolPerAsset,
```
QUESTION

No, blocker. But for follow-up. I would propose instead of adding all of that here, we should just trigger the message and the deposit call to the escrow directly FROM the `SyncRequyest` manager.

wischli replied:
But then the `BalanceSheet` is not the single source of truth for writing to the escrow anymore which I thought was one principe we wanted to follow. I suppose that changed given the latest work we did including this PR.

Would gladly let you do the follow up!

mustermeiszer reviewed (approved):
LGTM! 

I will do the follow up if this is fine?

wischli commented on src/vaults/BaseVaults.sol:
```diff
@@ -366,7 +373,8 @@ abstract contract BaseSyncDepositVault is BaseVault {
     function mint(uint256 shares, address receiver) public returns (uint256 assets) {
         assets = syncDepositManager.mint(address(this), shares, receiver, msg.sender);
-        SafeTransferLib.safeTransferFrom(asset, msg.sender, syncDepositManager.escrow(), assets);
+        // NOTE: For security reasons, transfer must stay at end of call
+        SafeTransferLib.safeTransferFrom(asset, msg.sender, address(_poolEscrowProvider.escrow(poolId)), assets);
```
Post mortem: Fixes https://github.com/Recon-Fuzz/centrifuge-review-protocol-v3/issues/36

GalloDaSballo replied:
LG, swaps the order to perform the unsafer call after
